crates/ruff_linter/src/rules/ruff/rules/helpers.rs

```diff
@@ -84,3 +87,72 @@ pub(super) fn is_descriptor_class(func: &Expr, semantic: &SemanticModel) -> bool
         })
     })
 }
+
+// fast check to disqualify any string literal that doesn't
+// have opening and closing brackets
+#[inline]
+fn has_brackets(possible_fstring: &str) -> bool {
+    let mut opening = false;
+    let mut closing = false;
+    for c in possible_fstring.chars() {
+        match c {
+            '{' => opening = true,
+            '}' => closing = true,
+            _ => {}
+        }
+    }
+    opening && closing
```

Comment:
We already rely on `memchr`, so you could consider using `memchr2` here to find all `{` and `}` characters, iterate over the matches, and ensure there's at least one of each? Though @BurntSushi would know if that will even be faster than what you have here :)

Reply:
Good suggestion!

Reply:
Some ideas:

1. Since you're just looking for ASCII bytes, I would jetison the use of `string.chars()`. That's doing UTF-8 decoding when you don't need it because UTF-8 is ASCII compatible. So just `for &byte in possible_fstring.as_bytes() {` will do.
2. When both `opening` and `closing` are `true`, the loop can stop. Although I could see this hurting more than helping.
3. `memchr2` should have substantially higher throughput than what you have here, but its latency is _probably_ worse. By how much, I'm not sure. What that means is that if this functions is usually called with very short strings (I mean very short, like a few bytes), then your code here might beat `memchr2`. But if they're usually longer, then `memchr2` is likely about the best you can do. (And to be clear, even if `memchr2` is faster here, I don't have enough context here to guess at whether it will be a holistic improvement.)